ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveMaterializedViewUtils.java

```diff
@@ -160,11 +181,79 @@ public static Boolean isOutdatedMaterializedView(
     return false;
   }
 
+  private static Boolean isOutdatedMaterializedView(
+          MaterializationSnapshot snapshot, Hive db,
+          Set<TableName> tablesUsed, Table materializedViewTable) throws HiveException {
+    List<String> tablesUsedNames = tablesUsed.stream()
+        .map(tableName -> TableName.getDbTable(tableName.getDb(), tableName.getTable()))
+        .collect(Collectors.toList());
+
+    Map<String, String> snapshotMap = snapshot.getTableSnapshots();
+    if (snapshotMap == null || snapshotMap.isEmpty()) {
+      LOG.debug("Materialized view " + materializedViewTable.getFullyQualifiedName() +
+              " ignored for rewriting as we could not obtain current snapshot ids");
+      return null;
+    }
+
+    Set<String> storedTablesUsed = materializedViewTable.getMVMetadata().getSourceTableFullNames();
+    for (String fullyQualifiedTableName : tablesUsedNames) {
+      // Note. If the materialized view does not contain a table that is contained in the query,
+      // we do not need to check whether that specific table is outdated or not. If a rewriting
+      // is produced in those cases, it is because that additional table is joined with the
+      // existing tables with an append-columns only join, i.e., PK-FK + not null.
+      if (!storedTablesUsed.contains(fullyQualifiedTableName)) {
+        continue;
+      }
+
+      Table table = db.getTable(fullyQualifiedTableName);
+      if (table.getStorageHandler() == null) {
+        LOG.debug("Materialized view {} ignored for rewriting as we could not storage handler of table {}",
+                materializedViewTable.getFullyQualifiedName(), fullyQualifiedTableName);
+        return null;
+      }
+      String currentTableSnapshot = table.getStorageHandler().getCurrentSnapshotId(table);
+      if (isBlank(currentTableSnapshot)) {
```

Comment:
The `currentTableSnapshot` will be never empty or null.

Reply:
It depends on the implementation: the default is `null`
HiveStorageHandler.java
```
  default String getCurrentSnapshotId(org.apache.hadoop.hive.ql.metadata.Table table) {
    return null;
  }
```

Reply:
According to the API contract
`@return String representation of the snapshotId or null if not supported.`
-1 is not a valid snapshotid. Maybe we should communicate this or the best would be to be in sync with the iceberg API to prevent confusion. If the table is empty return null snapshot.

Reply:
Refactored this API and it's usage:
* Renamed `getCurrentSnapshotId` to `getCurrentSnapshotContext` and return a `SnapshotContext` object wraps the `long snapshotId` instead of `String`
* `getCurrentSnapshotContext` default implementation returns `null` and the Iceberg implementation can also return `null` if the table is empty so no current snapshot exists.
* introduced `boolean areSnapshotsSupported` API method to distinguish between empty table and snapshots are not supported by the storage handler.
* Adjusted the usage at client side when checking MV is up-to-date.